### ISSUE TITLE
BACKUP to S3 bucket fails when using AUTH=implicit

### ISSUE THREAD

**Issue by cjireland** [cockroachdb/cockroach#56204]
Before looking at KMS encryption using v20.2, I wanted to check our existing backup/restore continues to work.  However, our backup/restore service starts throwing exceptions when we upgrade 20.1.6 -> 20.2.0-beta.3.  We get the same error when attempting to run a backup from command line (This works fine in 20.1.6.  The only thing changed was image version):
 
```
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
# Server version: CockroachDB CCL v20.2.0-beta.3 (x86_64-unknown-linux-gnu, built 2020/09/30 15:28:26, go1.13.14) (same version as client)
W201102 17:28:17.213547 1 cli/sql.go:1378  wide character support disabled
#
# Enter \? for a brief introduction.
#
root@:26257/defaultdb> BACKUP TO 's3://xxxxxxxxxxxxxxxxxxxx/adam-test?AUTH=implicit&AWS_REGION=eu-west-1'
                    -> AS OF SYSTEM TIME '2020-11-02T13:00:00' WITH revision_history;
ERROR: failed to get s3 object: WebIdentityErr: failed to retrieve credentials
caused by: RequestError: send request failed
caused by: Post https://sts.amazonaws.com/: dial tcp 54.239.29.25:443: i/o timeout
```
We use pod identity https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-technical-overview.html and the following get added by the AWS webhook:
```
   Environment:
      …
      AWS_REGION:                   eu-west-1
      AWS_ROLE_ARN:                 arn:aws:iam::…
      AWS_WEB_IDENTITY_TOKEN_FILE:  /var/run/secrets/eks.amazonaws.com/serviceaccount/token
```
I’ve checked the token and it is populated.

**Comment by pbardea** [comment]
I'm still investigating but here is my progress so far:
- It looks like the AUTH=implicit is still behaving correctly when tokens are stores in the `.aws/` config directory
- The difference in behavior was likely introduced by the AWS SDK version bump we did this release
    - Scanning through the release notes, 1.25.18 looks interesting: https://github.com/aws/aws-sdk-go/releases/tag/v1.25.18
    - It looks like the pod is failing to connect to `https://sts.amazonaws.com/` (the global endpoint for STS). My understanding is that if something like Amazon VPC is used, then it will fail to connect to endpoints outside its region.
    - From the release notes above, setting the environment variable `AWS_STS_REGIONAL_ENDPOINTS=regional` should make the client respect the region without the need for the endpoint parameter. This should ensure that the regional endpoint is used, rather than the global one.


This is where I am currently in my investigation. I'm still not sure why this behavior would have regressed since the behavior of the old SDK version should be the same behavior provided when the environment variable is set to `legacy`, which uses the global STS endpoint. My best guess here is to see if the SDK stopped looking at the `AWS_REGION` endpoint parameter all together and now only rely on the `AWS_STS_REGIONAL_ENDPOINTS` environment variable. This is my next line of investigation.

**Comment by pbardea** [comment]
One thing of note is that the AWS SDK [version](https://github.com/cockroachdb/vendored/tree/02478fc0f8362a6febb8b45c4d346ead61048298/github.com/aws/aws-sdk-go) (v1.16.19) that is used in 20.1.6, didn't support consuming web identity tokens. The earliest that I see it being supported is https://github.com/aws/aws-sdk-go/pull/2667, which was introduced in [v1.21.0](https://github.com/aws/aws-sdk-go/releases/tag/v1.21.0). 

I wasn't able to get AWS SSO fully set up today to try and repro with a Web Instance Token, but I would expect that it would ignore this flag in 20.1.6, based on my current reading of the code. I wonder if with the 20.1.6 version of the SDK, it was just ignoring this ENV variable and was able to fallback to some environment keys, but now with the new SDK version it tries to use the web instance token but fails to connect.

I did try to use a assume a role in hopes to hit `https://sts.amazonaws.com` that way, but unexpectedly ran into `InvalidAction: Could not find operation AssumeRole for version 2011-0615`... so failed to make much progress pulling that thread.

**Comment by cjireland** [comment]
Hi @pbardea,

Been investigating this again now we are finished with the backup with passphrase work.

I am now using CRDB 20.2.0. 

I have set up a policy for AssumeRoleWithWebIdentity against the role we use. But no joy. I also tried adding AWS_STS_REGIONAL_ENDPOINTS=regional. When I did I got the message Error: fStatus code 403.

Digging a little further into here: https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-technical-overview.html I saw this link to here: https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html:

```
The containers in your pods must use an AWS SDK version that supports assuming an IAM role via an OIDC web identity token file. AWS SDKs that are included in Linux distribution package managers may not be new enough to support this feature. Be sure to use at least the minimum SDK versions listed below:

Java (Version 2) — 2.10.11
Java — 1.11.704
Go — 1.23.13
```

Which version of the AWS SDK does 20.2.x use?

Thanks.

**Comment by pbardea** [comment]
20.2 should be using `1.33.8` (https://github.com/cockroachdb/cockroach/blob/release-20.2/go.mod#L27).

**Comment by cjireland** [comment]
Note, the containers do not run as root but they have `fsGroup 1000`.
I observed that  /var/run/secrets/eks.amazonaws.com/serviceaccount/token is owned by root BUT I can see the contents if I cat it.

**Comment by cjireland** [comment]
@pbardea is there any further logging I can enable/log I can check, to see what's actually happening here?

I am back at the `fStatus code 403, request_id xxxxxx host_id yyyyyyyy `error. 

Thanks!

**Comment by pbardea** [comment]
The first thing to comes in mind is to make sure that the file it is trying to access is the one we expect, then to investigate why we're getting a forbidden code response.

Is that the full error message? A full error message might help pinpoint exactly what is making the request to confirm the resource that we're attempting to access with that request.

**Comment by cjireland** [comment]
Hi @pbardea,

OK we have cracked it. 

I tried a sidecar container with awscli and that had problems writing to the S3 bucket as well.

In the end the solution was:

1. Add two environment variables to the cockroach pod:
```
AWS_DEFAULT_REGION=eu-west-1
AWS_STS_REGIONAL_ENDPOINTS=regional
```

2. Modify the permissions on the S3 bucket to allow both s3 (GetObject,ListBucket,PutObject) and kms permissions (KmsEncrypt, KmsDecrypt, etc.).

Thanks for all your help!
